lib/chat/key_share.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Chat.KeyShare do
+  @moduledoc "Manipulate social sharing keys"
+
+  alias Chat.Identity
+  alias Chat.{ChunkedFiles, ChunkedFilesMultisecret}
+  alias Chat.Upload.UploadKey
+  alias Chat.{Dialogs, Dialogs.Dialog, Messages}
+
+  alias Phoenix.PubSub
+
+  @threshold 4
+
+  def generate_key_shares({%Identity{private_key: private_key} = _me, users}) do
+    shares =
+      private_key
+      |> Enigma.hide_secret_in_shares(users |> Enum.count(), @threshold)
+
+    share_count = Enum.count(shares)
+
+    for i <- 0..(share_count - 1), into: [] do
+      %{
+        user: Enum.at(users, i),
+        key: Enum.at(shares, i) |> Base.encode64()
+      }
+    end
+  end
+
+  def send_shares(shares, {me, time_offset}) do
+    shares
+    |> Enum.each(fn share ->
+      with dialog <- Dialogs.find_or_open(me, share.user),
+           file_info <- %{
+             size: byte_size(share.key),
+             time: Chat.Time.monotonic_to_unix(time_offset)
+           },
+           entry <- entry(file_info, me),
+           destination <- destination(dialog),
+           file_key <- UploadKey.new(destination, dialog.b_key, entry),
+           file_secret <- ChunkedFiles.new_upload(file_key) do
+        save({file_key, share.key}, {file_info.size, file_secret})
+        send(entry, dialog, me, {file_key, file_secret, file_info.time})
+      end
+    end)
+  end
+
+  def schema, do: %{users: {:array, :string}}
```

Comment:
what is this for ?

Reply:
it moved to KeyShareForm, use such schema to validate share users form, removed here